traffic_ops/testing/api/v4/deliveryservices_test.go

```diff
@@ -98,6 +99,64 @@ func TestDeliveryServices(t *testing.T) {
 	})
 }
 
+func CreateTestDeliveryServiceWithGeoLimitCountries(t *testing.T) {
+
+	if len(testData.DeliveryServices) == 0 {
+		t.Fatalf("no deliveryservices to run the test on, quitting")
+	}
+
+	cdn := createBlankCDN("geoLimitCDN", t)
+	opts := client.NewRequestOptions()
+	opts.QueryParameters.Set("name", "HTTP")
+	types, _, err := TOSession.GetTypes(opts)
+	if err != nil {
+		t.Fatalf("unable to get Types: %v - alerts: %+v", err, types.Alerts)
+	}
+	if len(types.Response) < 1 {
+		t.Fatal("expected at least one type")
+	}
+	customDS := getCustomDS(cdn.ID, types.Response[0].ID, "geo-limit-countries-test-ds-name", "edge", "https://test-geo-limit.com", "geo-limit-countries-test-ds-xml-id")
+	customDS.Protocol = util.IntPtr(0)
+	customDS.GeoLimit = util.IntPtr(2)
+	//geoLimitCountries := []string{"US   ", "CA"}
+	customDS.GeoLimitCountries = []string{"US   ", "CA"}
+
+	resp, _, err := TOSession.CreateDeliveryService(customDS, client.RequestOptions{})
+	if err != nil {
+		t.Errorf("expected no error while creating a new ds, but got %v", err)
+	}
+	if len(resp.Response) != 1 {
+		t.Fatalf("expected 1 response in return of a create DS request, but got %d", len(resp.Response))
+	}
+	if resp.Response[0].GeoLimitCountries == nil {
+		t.Fatalf("got nothing in geo limit countries")
+	}
+	arr := ([]string)(resp.Response[0].GeoLimitCountries)
+	if len(arr) != 2 || arr[0] != "US" || arr[1] != "CA" {
+		t.Errorf("expected geo limit countries: US,CA; actual: %s", arr)
+	}
+	opts = client.NewRequestOptions()
+	opts.QueryParameters.Set("xmlId", *customDS.XMLID)
+	deliveryServices, _, err := TOSession.GetDeliveryServices(opts)
+	if err != nil {
+		t.Fatalf("couldn't get ds: %v", err)
+	}
+	if len(deliveryServices.Response) != 1 {
+		t.Fatal("couldn't get exactly one ds in the response, quitting")
+	}
+	dsID := deliveryServices.Response[0].ID
+	customDS.GeoLimitCountries = []string{"US   ", "CA", "12"}
+	_, _, err = TOSession.UpdateDeliveryService(*dsID, customDS, client.RequestOptions{})
```

Comment:
This line (and L153) will segfault if Traffic Ops erroneously returns a Delivery Service with a null or omitted ID